alf/networks/initializers.py

```diff
@@ -19,16 +19,49 @@
 import torch
 import torch.nn as nn
 
+import alf.utils.math_ops as math_ops
 
-def _calculate_gain(nonlinearity, nonlinearity_param):
-    """Handle extra activations.
-    TODO: might compute numerical gain instead.
+
+@gin.configurable
+def _numerical_calculate_gain(nonlinearity, num_samples=1000):
+    """Compute the gain in a numerical way by sampling variables. For the sampled
+    variables, we compute the sqrt inverse of their average second moment, and
+    use it as the suggested gain.
+
+    Assume y is the output, w is the weight (mean=0), and x is the input, then
+
+    Var(y) = Var(w) * E(x^2)
+
+    Args:
+        nonlinearity (Callable): any callable activation function
+        num_samples (int): number of samples for approximating the gain
+
+    Returns:
+        gain (float): a gain factor that will be applied to the init weights
+    """
+    tensor = torch.randn(num_samples)
```

Comment:
we don't need to resort to sampling for doing this as its error is proportional to 1/sqrt(num_samples). We can do numerical integration:
```python
x=torch.arange(-r,r,delta)
variance = pdf(x)*(nonlinearity(x)**2)).sum()*delta
```

Reply:
Changed, but the result seems almost no difference.